Supporting TTL update at the frontend

vgkholla opened the pull request:
This change plugs in all the changes required to support TTL updates at the frontend

cgtz commented on ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java:
```diff
@@ -246,11 +250,32 @@ public void handlePost(RestRequest restRequest, RestResponseChannel restResponse
+      logger.trace("Handling PUT request - {}", restRequest.getUri());
+      checkAvailable();
+      // TODO: make this non blocking once all handling of individual methods is moved to their own classes
+      securityService.preProcessRequest(restRequest).get();
```
should this call happen inside of ttlUpdateHandler?

vgkholla replied:
I'm not sure - I saw that you did this in the `PostBlobHandler` but it seems like the preprocessing is actually not related to the kind of request - I'm thinking that we can unify the code as one function is `AmbryBlobStorageService` - what do you think?

cgtz commented on ambry-frontend/src/main/java/com.github.ambry.frontend/AmbrySecurityService.java:
```diff
@@ -101,7 +104,7 @@ public void postProcessRequest(RestRequest restRequest, Callback<Void> callback)
     }
     // check preconditions for DELETE request
-    if (restRequest.getRestMethod() == RestMethod.DELETE) {
+    if (restRequest.getRestMethod() == RestMethod.DELETE || restRequest.getRestMethod().equals(RestMethod.PUT)) {
```
minor: you can use == here since it's an enum (or change the `== RestMethod.DELETE` to `.equals(RestMethod.DELETE)` for consistency)

vgkholla replied:
Done

cgtz commented on ambry-frontend/src/main/java/com.github.ambry.frontend/AmbrySecurityService.java:
```diff
@@ -130,7 +133,8 @@ public void processResponse(RestRequest restRequest, RestResponseChannel respons
-      if (blobInfo == null && !restRequest.getRestMethod().equals(RestMethod.OPTIONS)) {
+      if (blobInfo == null && !restRequest.getRestMethod().equals(RestMethod.OPTIONS) && !restRequest.getRestMethod()
+          .equals(RestMethod.PUT)) {
         if (!operationOrBlobId.equals(Operations.GET_SIGNED_URL)) {
```
Kind of orthogonal, but I believe we ignore the URI (for now) on POST requests so this check wouldn't trigger if someone attempted a `POST /signedUrl`

vgkholla replied:
but that's not a valid op - only `GET /signedUrl` is. Something I'm missing?

cgtz replied:
Yep, it's not a valid operation. I don't think there is any code in `PostBlobHandler` checking the path on POST requests, so someone could send a `POST /signedUrl` request (and it would be handled the same way as a `POST /` request. However, when the code gets to this `blobInfo != null` check will not get triggered. This isn't a huge problem since the code will just throw a NPE a few lines down. 

Anyways, `AmbryBlobStorageService` is probably the place to check that the path on a POST request is `/`

cgtz commented on ambry-frontend/src/main/java/com.github.ambry.frontend/TtlUpdateHandler.java:
```diff
@@ -0,0 +1,261 @@
+      metrics.updateBlobTtlSecurityPostProcessRequestTimeInMs.update(processingStartTimeMs - operationStartTimeMs);
+      try {
+        if (exception == null) {
+          String serviceId = RestUtils.getHeader(restRequest.getArgs(), RestUtils.Headers.SERVICE_ID, true);
```
I guess we are requiring service-id here since this is a new API? I think it was optional on delete.

vgkholla replied:
Right - I think we should require it in everything but left didn't make that change for now

cgtz commented on ambry-frontend/src/main/java/com.github.ambry.frontend/TtlUpdateHandler.java:
```diff
@@ -0,0 +1,261 @@
+          LOGGER.debug("Updated TTL of {}",
+              RestUtils.getHeader(restRequest.getArgs(), RestUtils.Headers.BLOB_ID, true));
+          restResponseChannel.setHeader(RestUtils.Headers.DATE, new GregorianCalendar().getTime());
+          restResponseChannel.setHeader(RestUtils.Headers.CONTENT_LENGTH, 0);
```
is setting the content length header here required, or will `submitResponse` do that if you don't give it a response body?

vgkholla replied:
It doesn't but that is probably something to think about - I don't want to do it in this patch though - not sure if there are implications

cgtz commented on ambry-frontend/src/main/java/com.github.ambry.frontend/TtlUpdateHandler.java:
```diff
@@ -0,0 +1,261 @@
+        if (exception == null) {
+          LOGGER.debug("Updated TTL of {}",
+              RestUtils.getHeader(restRequest.getArgs(), RestUtils.Headers.BLOB_ID, true));
+          restResponseChannel.setHeader(RestUtils.Headers.DATE, new GregorianCalendar().getTime());
```
I think security service might also be setting the date header

vgkholla replied:
Unable to find the one in `AmbrySecurityService` - will chat with you

vgkholla commented on ambry-frontend/src/main/java/com.github.ambry.frontend/AmbrySecurityService.java:
```diff
@@ -101,7 +104,7 @@ public void postProcessRequest(RestRequest restRequest, Callback<Void> callback)
       throw new IllegalArgumentException("RestRequest or Callback is null");
     }
     // check preconditions for DELETE request
```
Done

vgkholla commented on ambry-frontend/src/main/java/com.github.ambry.frontend/FrontendMetrics.java:
```diff
@@ -192,6 +207,10 @@ public FrontendMetrics(MetricRegistry metricRegistry) {
+    // PUT
+    updateBlobTtlMetrics = new RestRequestMetrics(GetSignedUrlHandler.class, "UpdateBlobTtl", metricRegistry);
+    updateBlobTtlSSLMetrics =
+        new RestRequestMetrics(GetSignedUrlHandler.class, "UpdateBlobTtl" + SSL_SUFFIX, metricRegistry);
```
Done

vgkholla commented on ambry-frontend/src/main/java/com.github.ambry.frontend/FrontendMetrics.java:
```diff
@@ -192,6 +207,10 @@ public FrontendMetrics(MetricRegistry metricRegistry) {
     optionsMetrics = new RestRequestMetrics(AmbryBlobStorageService.class, "Options", metricRegistry);
     optionsSSLMetrics = new RestRequestMetrics(AmbryBlobStorageService.class, "Options" + SSL_SUFFIX, metricRegistry);
+    // PUT
+    updateBlobTtlMetrics = new RestRequestMetrics(GetSignedUrlHandler.class, "UpdateBlobTtl", metricRegistry);
```
Fixed

vgkholla commented on ambry-frontend/src/test/java/com.github.ambry.frontend/AmbryBlobStorageServiceTest.java:
```diff
@@ -1584,18 +1541,82 @@ private void verifyUserMetadataHeaders(JSONObject expectedHeaders, MockRestRespo
+  }
+
   /**
    * Deletes the blob with blob ID {@code blobId} and verifies the response returned.
```
Done

vgkholla commented on ambry-router/src/test/java/com.github.ambry.router/InMemoryRouter.java:
```diff
@@ -319,16 +327,21 @@ public void close() {
       completeOperation(futureResult, callback, null,
           new RouterException("Cannot accept operation because Router is closed", RouterErrorCode.RouterClosed));
     } else if (verifiableProperties.containsKey(OPERATION_THROW_EARLY_RUNTIME_EXCEPTION)) {
       throw new RuntimeException(OPERATION_THROW_EARLY_RUNTIME_EXCEPTION);
```
No, because it throws immediately. That is an indication that op should not be continued. Also since this is an exception, the value of `continueOp` doesn't matter because it won't be returned

vgkholla commented on ambry-frontend/src/test/java/com.github.ambry.frontend/TtlUpdateHandlerTest.java:
```diff
@@ -0,0 +1,289 @@
+import static org.junit.Assert.*;
+
+
+public class TtlUpdateHandlerTest {
```
Added a line

vgkholla commented on ambry-frontend/src/test/java/com.github.ambry.frontend/TtlUpdateHandlerTest.java:
```diff
@@ -0,0 +1,289 @@
+
+  /**
+   * Verifies that the TTL is {@code expectedTtl}
+   * @param expectedTtl the expected TTL of the blob
```
Added units

vgkholla commented on ambry-frontend/src/test/java/com.github.ambry.frontend/TtlUpdateHandlerTest.java:
```diff
@@ -0,0 +1,289 @@
+   * @param expectedTtl the expected TTL of the blob
+   * @throws Exception
+   */
+  private void assertTtl(long expectedTtl) throws Exception {
```
Done

vgkholla commented on ambry-rest/src/main/java/com.github.ambry.rest/NettyResponseChannel.java:
```diff
@@ -410,14 +410,31 @@ private FullHttpResponse getErrorResponse(Throwable cause) {
   }
 
+  /**
+   * @param status the {@link HttpResponseStatus}
```
Nothing more to describe than what is in the `@return`. Anything you think I can add in a description?

vgkholla commented on ambry-frontend/src/main/java/com.github.ambry.frontend/TtlUpdateHandler.java:
```diff
@@ -0,0 +1,261 @@
+   * Callback for {@link SecurityService#processRequest(RestRequest, Callback)} that subsequently issues a call to
+   * the {@link IdConverter}
+   */
+  private class SecurityProcessRequestCallback implements Callback<Void> {
```
Will refactor to use new code

vgkholla commented:
@cgtz @jsjtzyy Addressed comments. Also,
1. Refacored to use chain callbacks
2. Added 'allow' header when the response will be 405 (as required by RFC)
3. New tests for new functionality

jsjtzyy commented on ambry-frontend/src/test/java/com.github.ambry.frontend/TtlUpdateHandlerTest.java:
```diff
@@ -0,0 +1,297 @@
+  private static final Container REF_CONTAINER = REF_ACCOUNT.getContainerById(Container.DEFAULT_PRIVATE_CONTAINER_ID);
+  private static final ClusterMap CLUSTER_MAP;
+  private static final String SERVICE_ID = "TtlUpdateHandlerTest";
+  private static final long TTL_SECS = 7200;
```
There is `TTL_SECS` in a bunch of test files.  Maybe we can add that into a TestUtil file and thus `static` makes more sense.

vgkholla replied:
Added to `com.github.ambry.utils.TestUtils`

jsjtzyy commented on ambry-rest/src/test/java/com.github.ambry.rest/RestTestUtils.java:
```diff
@@ -51,6 +55,22 @@ public static HttpRequest createRequest(HttpMethod httpMethod, String uri, HttpH
+   * @param content the content to add to the request
+   * @return a {@link FullHttpRequest} with the given parameters.
+   */
+  public static FullHttpRequest createRequest(HttpMethod httpMethod, String uri, HttpHeaders headers, byte[] content) {
```
Is it better to call the method `createFullRequest` to distinguish it from another `createRequest` method in this file?

vgkholla replied:
Done

vgkholla commented on ambry-api/src/main/java/com.github.ambry/rest/RestUtils.java:
```diff
@@ -185,6 +185,10 @@
+    /**
+     * {@code "allow"}
+     */
+    public final static String ALLOW = "allow";
```
According to [RFC7231|https://tools.ietf.org/html/rfc7231#section-6.5.5], the `Allow` header must be returned if the status is `405 Method Not Allowed`. TTL update can return this status in some cases. Does that answer your question?

vgkholla commented on ambry-frontend/src/test/java/com.github.ambry.frontend/AmbryBlobStorageServiceTest.java:
```diff
@@ -103,6 +103,8 @@
  * Unit tests for {@link AmbryBlobStorageService}. Also tests {@link AccountAndContainerInjector}.
  */
 public class AmbryBlobStorageServiceTest {
+  private static final long TTL_SECS = 7200;
```
Since it is actually a constant that does not depend on different instances, it is appropriate that it is `static`

vgkholla commented on ambry-frontend/src/test/java/com.github.ambry.frontend/AmbryBlobStorageServiceTest.java:
```diff
@@ -1158,7 +1159,7 @@ private void doRuntimeExceptionRouterTest(RestMethod restMethod) throws Exceptio
+  // postGetHeadUpdateDeleteTest() helpers
 
   /**
    * Tests blob POST, GET, HEAD and DELETE operations on the given {@code container}.
```
Done

vgkholla commented:
@jsjtzyy comments addressed

lightningrob commented on ambry-api/src/test/java/com.github.ambry/rest/MockRestResponseChannel.java:
```diff
@@ -132,8 +132,12 @@ public synchronized void onResponseComplete(Exception exception) {
+          // clear headers except for the value of Allow
+          String allow = getHeader(RestUtils.Headers.ALLOW);
           responseMetadata.put(RESPONSE_HEADERS_KEY, new JSONObject());
+          if (allow != null && !allow.isEmpty()) {
```
Minor: if we don't import guava which has nice utilities like Strings.isNullOrEmpty(), then would be good to implement our own somewhere.

vgkholla replied:
Done

lightningrob commented on ambry-frontend/src/main/java/com.github.ambry.frontend/AmbrySecurityService.java:
```diff
@@ -130,7 +133,8 @@ public void processResponse(RestRequest restRequest, RestResponseChannel respons
         operationOrBlobId = operationOrBlobId.substring(1);
       }
-      if (blobInfo == null && !restRequest.getRestMethod().equals(RestMethod.OPTIONS)) {
+      if (blobInfo == null && !restRequest.getRestMethod().equals(RestMethod.OPTIONS) && !restRequest.getRestMethod()
```
Minor: I'd move up the line that extracts restMethod into a variable and use that here.

cgtz commented on ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java:
```diff
@@ -213,11 +220,41 @@ public void handlePost(RestRequest restRequest, RestResponseChannel restResponse
+    } catch (Exception e) {
+      submitResponse(restRequest, restResponseChannel, null, extractExecutionExceptionCause(e));
+    } finally {
+      frontendMetrics.putPreProcessingTimeInMs.update(preProcessingTime);
```
could we make one `submitResponse` call in this finally block? i.e. reassign the variable `e` in the callback logic and call `submitResponse(..., null, e)` in `finally`

vgkholla commented on ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java:
```diff
@@ -213,11 +220,41 @@ public void handlePost(RestRequest restRequest, RestResponseChannel restResponse
+              && ((RouterException) e).getErrorCode() == RouterErrorCode.BlobUpdateNotAllowed) {
+            try {
+              restResponseChannel.setHeader(Headers.ALLOW, TTL_UPDATE_REJECTED_ALLOW_HEADER_VALUE);
+            } catch (RestServiceException exp) {
```
Good point - done
